Enforce and default Kafka configuration

nasark opened the pull request:
- Enforce Kafka by making the orchestrator wait for Kafka service to be ready (requires Kafka network policy)
- default `deployMessagingService` to true from CR as Kafka will be the default configuration -> this will retain the option to disable Kafka in case things go wrong


Depends on:
- [x] https://github.com/ManageIQ/manageiq-pods/pull/935
- [x] https://github.com/ManageIQ/manageiq-pods/pull/936

Needed for:
- [ ] https://github.com/ManageIQ/manageiq-pods/pull/1005

Ref:
- https://github.com/ManageIQ/manageiq/issues/22225

@miq-bot add_label enhancement
@miq-bot assign @Fryguy 
@miq-bot add_reviewer @agrare

Fryguy commented:
@bdunne Please review.

Fryguy commented on manageiq-operator/api/v1alpha1/manageiq_types.go:
```diff
@@ -63,10 +63,6 @@ type ManageIQSpec struct {
 	// +optional
 	DatabaseVolumeCapacity string `json:"databaseVolumeCapacity,omitempty"`
 
-	// Flag to indicate if Kafka and Zookeeper should be deployed (default: false)
-	// +optional
-	DeployMessagingService *bool `json:"deployMessagingService,omitempty"`
```
Was this ever shipped? or was this transient?  We may need to keep it since it might break backward compat.

nasark replied:
I see, would it better to keep the `deployMessagingService` option and set its default value to `true` as opposed to the current `false` then?

agrare commented:
@nasark so this removes the option to deploy kafka, does this actually do the deployment?

nasark commented:
> @nasark so this removes the option to deploy kafka, does this actually do the deployment?

Actually this removes the flag that would disable Kafka by default. Removing the flag altogether makes it so that Kafka is deployed alongside the other services by default.

agrare commented:
Awesome thanks @nasark can you tell what it deploys is it ManageIQ/container-kafka ?

nasark commented:
If removing the flag altogether is a concern for backwards compatibility, then we could just set the default value for `deployMessagingService` to `true` as well. Some clarification and/or elaboration regarding https://github.com/ManageIQ/manageiq-pods/pull/896#discussion_r995099738 would be appreciated @Fryguy

nasark commented:
> Awesome thanks @nasark can you tell what it deploys is it ManageIQ/container-kafka ?

The resources deployed are handled here https://github.com/ManageIQ/manageiq-pods/blob/30480277dff0accfb679bef93b9d206257f7e0f3/manageiq-operator/controllers/manageiq_controller.go#L364-L429

Fryguy commented:
> If removing the flag altogether is a concern for backwards compatibility, then we could just set the default value for `deployMessagingService` to `true` as well. Some clarification/elaboration regarding [#896 (comment)](https://github.com/ManageIQ/manageiq-pods/pull/896#discussion_r995099738) would be appreciated @Fryguy

We may also need a CR migration, but it might not matter if we just don't honor it?  cc @bdunne

Fryguy commented:
@agrare That would explain why that image hasn't been updated in years 😆

nasark commented:
Returning to this, perhaps it is easier to just default `deployMessagingService` to `true` instead of removing it from the CR altogether. Let me know what you think and how we can proceed with this @Fryguy @bdunne

nasark commented:
I've split some of this up and moved into other PRs based on issues I have found in my testing:
- https://github.com/ManageIQ/manageiq-pods/pull/935
- https://github.com/ManageIQ/manageiq-pods/pull/936

Fryguy commented on manageiq-operator/api/v1alpha1/helpers/miq-components/network_policies.go:
```diff
@@ -217,44 +217,6 @@ func NetworkPolicyAllowPostgres(cr *miqv1alpha1.ManageIQ, scheme *runtime.Scheme
 	return networkPolicy, f
 }
 
-func NetworkPolicyAllowKafka(cr *miqv1alpha1.ManageIQ, scheme *runtime.Scheme, c *client.Client) (*networkingv1.NetworkPolicy, controllerutil.MutateFn) {
```
Why does this go away?

nasark commented:
@miq-bot remove_label stale

nasark commented:
@miq-bot remove_label stale

nasark commented:
@miq-bot add_label quinteros/yes?

bdunne commented on manageiq-operator/api/v1alpha1/manageiq_types.go:
```diff
@@ -64,8 +64,7 @@ type ManageIQSpec struct {
 	DatabaseVolumeCapacity string `json:"databaseVolumeCapacity,omitempty"`
 
-	// Flag to indicate if Kafka and Zookeeper should be deployed (default: false)
-	// +optional
```
I think we need to keep this comment

nasark replied:
I've removed the deprecated bit, do you want to keep the `optional` comment too? Technically it will be used by default due to the `true` value but don't think that makes it optional?

bdunne commented on manageiq-operator/api/v1alpha1/manageiq_types.go:
```diff
@@ -64,8 +64,7 @@ type ManageIQSpec struct {
 
-	// Flag to indicate if Kafka and Zookeeper should be deployed (default: false)
-	// +optional
+	// Flag to indicate if Kafka and Zookeeper should be deployed (default: true)
```
The deprecated part was fine.  We need to keep the `+optional` otherwise the CR validation will require users to fill in that field.  See `+optional` in https://book.kubebuilder.io/reference/markers/crd-validation
```suggestion
	// Deprecated: Flag to indicate if Kafka and Zookeeper should be deployed (default: true)
	// +optional
```

bdunne commented on manageiq-operator/api/v1alpha1/helpers/miq-components/cr.go:
```diff
@@ -51,7 +51,7 @@ func databaseVolumeCapacity(cr *miqv1alpha1.ManageIQ) string {
 func deployMessagingService(cr *miqv1alpha1.ManageIQ) bool {
 	if cr.Spec.DeployMessagingService == nil {
-		return false
+		return true
```
I think we can remove the conditional completely and always return true unless we add some conditional logic to the service checking in the entrypoint script.  As is it, the orchestrator will wait foerver for the kafka if the user sets this to false, right?

Fryguy commented:
Backported to `quinteros` in commit fe79478dc8a65610d2743801b4973825fffe5985.

```text
commit fe79478dc8a65610d2743801b4973825fffe5985
Author: Brandon Dunne <brandondunne@hotmail.com>
Date:   Wed Jan 3 12:06:32 2024 -0500

    Merge pull request #896 from nasark/require_kafka
    
    Enforce and default Kafka configuration
    
    (cherry picked from commit 09da4c7f3083fad1b56f345d7e6e8c0063a37f58)
```
